### ISSUE TITLE
Does scaled softmax of qkv perform scaling operations duplicately?

### ISSUE THREAD

**Issue by marigoold** [NVIDIA/Megatron-LM#175]
I implemented the t5 model and compared it with the implementation in Megatron using the same weights, with the following results
```
max embedding diff:  0.0
max encoder layers0 diff:  0.00013916194
max encoder layers1 diff:  0.0049294047
max encoder layers2 diff:  0.0081669465
max encoder layers3 diff:  0.011228945
max encoder layers4 diff:  0.014814734
max encoder layers5 diff:  0.015128866
max encoder layers6 diff:  0.016686544
max encoder layers7 diff:  0.019047052
max encoder layers8 diff:  0.022120267
max encoder layers9 diff:  0.023271054
max encoder layers10 diff:  0.023713648
max encoder layers11 diff:  0.025910359
```

When I set the `layer_index` of all encoder layers in my implementation to 0, the comparison results are as follows: 
```
max embedding diff:  0.0
max encoder layers0 diff:  0.00013916194
max encoder layers1 diff:  0.00024357438
max encoder layers2 diff:  0.0003227368
max encoder layers3 diff:  0.00041988492
max encoder layers4 diff:  0.00054493546
max encoder layers5 diff:  0.0006237328
max encoder layers6 diff:  0.0006262362
max encoder layers7 diff:  0.0006904714
max encoder layers8 diff:  0.0007108152
max encoder layers9 diff:  0.0008518249
max encoder layers10 diff:  0.0008100569
max encoder layers11 diff:  0.0008664727
```

I tracked and debugged the Megatron t5 model and found that both the numerator and denominator of attention_score were scaled by layer_index before calculating the softmax.

numerator scaling: 
https://github.com/NVIDIA/Megatron-LM/blob/9a8b89acd8f6ba096860170d0e30ddc0bc2bacd4/megatron/model/fused_softmax.py#L171-L176

denominator scaling:
https://github.com/NVIDIA/Megatron-LM/blob/9a8b89acd8f6ba096860170d0e30ddc0bc2bacd4/megatron/model/transformer.py#L156-L159

The two scaling operations cancel each other out and the result is equivalent to that with no scaling, therefore the scaling factor of each layer of encoder is equal.

**Comment by RuiWang1998** [comment]
I believe this is because the qk scale here mostly deals with numerical stability of fp16, where, if divided, logits are easier to represent and scaled back before softmax. This is semantically different than the heuristics that the attention scores should be scaled based on layer indices.
